store.go

```diff
@@ -3815,3 +3834,56 @@ func (s *store) GarbageCollect() error {
 
 	return firstErr
 }
+
+// List returns a ListResult structure that contains layer, image, or container
+// extracts according to the values in ListOptions.
+func (s *store) MultiList(options StoreListOptions) (StoreListResult, error) {
+	// TODO: Possible optimization: Deduplicate content from multiple stores.
+	out := StoreListResult{}
+
+	layerStores, err := s.allLayerStores()
+	if err != nil {
+		return out, err
+	}
+
+	if options.Layers {
+		for _, roStore := range layerStores {
+			if err := roStore.startReading(); err != nil {
+				return out, err
+			}
+			defer roStore.stopReading()
+			layers, err := roStore.Layers()
+			if err != nil {
+				return out, err
+			}
+			out.Layers = append(out.Layers, layers...)
+		}
+	}
+
+	if options.Images {
+		for _, roStore := range s.allImageStores() {
+			if err := roStore.startReading(); err != nil {
+				return out, err
+			}
+			defer roStore.stopReading()
+
+			images, err := roStore.Images()
+			if err != nil {
+				return out, err
+			}
+			out.Images = append(out.Images, images...)
+		}
+	}
+
+	if options.Containers {
+		containers, _, err := readContainerStore(s, func() ([]Container, bool, error) {
+			res, err := s.containerStore.Containers()
+			return res, true, err
+		})
+		if err != nil {
+			return out, err
```

Comment:
Throughout: On error paths, it’s better to explicitly return a `…Result{}`, so that callers don’t get any ideas about using the value anyway.